### PR TITLE
WavefrontHistogram for Dropwizard metrics 3.x/4.x

### PR DESCRIPTION
**Pull request by hanwavefront**
[MONIT-9992](https://wavefront.atlassian.net/browse/MONIT-9992)

### CODE REVIEW

File: java-lib/src/main/java/com/codahale/metrics/WavefrontHistogram.java [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+    if (means != null && counts != null) {
+      int n = Math.min(means.size(), counts.size());
+      MinuteBin current = getCurrent();
+      for (int i = 0; i < n; ++i) {
+        current.getDist().add(means.get(i), counts.get(i));
+      }
+    }
+  }
+
+  public void update(double value) { getCurrent().getDist().add(value); }
```
Comment (vikramraman) [inline java-lib/src/main/java/com/codahale/metrics/WavefrontHistogram.java]: Break into new lines similar to `getCount()`.

File: java-lib/src/main/java/com/codahale/metrics/WavefrontHistogram.java [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+
+  @Override
+  public void update(long value) { update((double)value); }
```
Comment (vikramraman) [inline java-lib/src/main/java/com/codahale/metrics/WavefrontHistogram.java]: Ditto. New lines.

File: java-client/src/main/java/com/wavefront/integrations/WavefrontDirectSender.java [hunk 3]
```diff
@@ -99,6 +149,60 @@ static String pointToString(String name, double value, @Nullable Long timestamp,
     }
     sb.append("source=").append(quote).append(escapeQuotes(source)).append(quote);
 
+    sb.append(pointTagsToString(pointTags));
+    return sb.toString();
+  }
+
+  @Nullable
+  static String histogramToString(HistogramGranularity histogramGranularity, @Nullable Long timestamp,
```
Comment (vikramraman) [inline java-client/src/main/java/com/wavefront/integrations/WavefrontDirectSender.java]: Let's create a HistogramConstants.java under https://github.com/wavefrontHQ/java/tree/b9c5258ff7bf729cf4af82af5ca55c9850f49e16/java-lib/src/main/java/com/wavefront/common and put this logic there and re-use here and Wavefront.java.

Have the common code throw an IllegalArgumentException but catch and log here and use as is within Wavefront.java.

File: java-client/src/main/java/com/wavefront/integrations/WavefrontDirectSender.java [hunk 4]
```diff
@@ -17,6 +19,9 @@
 import javax.validation.constraints.NotNull;
 import javax.ws.rs.core.Response;
 
+import static com.wavefront.api.agent.Constants.PUSH_FORMAT_GRAPHITE_V2;
+import static com.wavefront.api.agent.Constants.PUSH_FORMAT_HISTOGRAM;
+
```
Comment (vikramraman) [inline java-client/src/main/java/com/wavefront/integrations/WavefrontDirectSender.java]: Add another author tag with your name as well.

File: java-client/src/main/java/com/wavefront/integrations/WavefrontDirectSender.java [hunk 5]
```diff
@@ -111,26 +215,29 @@ static String pointToString(String name, double value, @Nullable Long timestamp,
...
     if (!isConnected()) {
         return;
     }
+    internalFlush(pointBuffer, PUSH_FORMAT_GRAPHITE_V2);
+    internalFlush(histogramBuffer, PUSH_FORMAT_HISTOGRAM);
+  }
 
-    List<String> points = getPointsBatch();
+  private void internalFlush(LinkedBlockingQueue<String> buffer, String format) throws IOException {
+    List<String> points = getBatch(buffer);
```
Comment (vikramraman) [inline java-client/src/main/java/com/wavefront/integrations/WavefrontDirectSender.java]: We currently only flush up to BATCH_SIZE (10k) points per flush. I am wondering whether we should change getBatch to return a `List<List<String>>`. Basically split into batches and flush the entire buffer per flush. Something to run by @panghy .

File: java-client/src/main/java/com/wavefront/integrations/WavefrontDirectSender.java [hunk 6]
```diff
@@ -70,14 +80,54 @@ public void send(String name, double value, @Nullable Long timestamp, String sou
     addPoint(name, value, timestamp, source, pointTags);
   }
 
+  @Override
+  public void send(HistogramGranularity histogramGranularity, List<Pair<Double, Integer>> distribution, String name) throws IOException {
```
Comment (vikramraman) [inline java-client/src/main/java/com/wavefront/integrations/WavefrontDirectSender.java]: Just noticed, the javadoc for several send methods say we will use the machine hostname if source is null, but we use DEFAULT_SOURCE. Perhaps we can cache and reuse the hostname, something like this:

```
private String getDefaultSource() throws UnknownHostException {
    if (defaultSource == null) {
      defaultSource = InetAddress.getLocalHost().getHostName();
    }
    return defaultSource;
  }
```

File: java-client/src/main/java/com/wavefront/integrations/Wavefront.java [hunk 7]
```diff
@@ -135,20 +166,65 @@ private void internalSend(String name, double value, @Nullable Long timestamp, S
...
+      return; // don't send if distribution is empty
+    }
+    if (!isConnected()) {
+      try {
+        connect();
+      } catch (IllegalStateException ex) {
+        // already connected.
+      }
+    }
+    if (histogramGranularity == null) {
```
Comment (vikramraman) [inline java-client/src/main/java/com/wavefront/integrations/Wavefront.java]: Refactor into common code and re-use. See comment in WavefrontDirectSender.

File: dropwizard-metrics/dropwizard-metrics/src/main/java/com/wavefront/integrations/metrics/WavefrontReporter.java [hunk 8]
```diff
@@ -1,27 +1,20 @@
 package com.wavefront.integrations.metrics;
 
+import com.codahale.metrics.*;
```
Comment (sushantdewan123) [inline dropwizard-metrics/dropwizard-metrics/src/main/java/com/wavefront/integrations/metrics/WavefrontReporter.java]: We actually avoid wild card imports.
You can copy system/.idea/codeStyleSettings.xml into java repo and IntelliJ will take care of this.

File: java-lib/src/main/java/com/yammer/metrics/core/WavefrontHistogram.java [hunk 9]
```diff
@@ -115,8 +116,8 @@ private MinuteBin getCurrent() {
     // bins with clear == true flag will drain (CONSUMER) the list,
     // so synchronize the access to the respective 'bins' list
     synchronized (bins) {
-      if (bins.isEmpty() || bins.getLast().minMillis != currMinMillis) {
-        bins.add(new MinuteBin(currMinMillis));
+      if (bins.isEmpty() || bins.getLast().getMinuteMillis() != currMinMillis) {
```
Comment (sushantdewan123) [inline java-lib/src/main/java/com/yammer/metrics/core/WavefrontHistogram.java]: Let's rename this to "currMinuteMillis" and elsewhere.
i.e. change minMillis to minuteMillis across the board.

File: java-lib/src/main/java/com/codahale/metrics/WavefrontHistogram.java [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+  /**
+   * Cache of thread ids to bin queues. By giving each thread its own bin queue, we can ensure
+   * thread safety by locking only the relevant bin queue for a particular thread. This is more performant
+   * than locking the entire histogram.
+   *
+   * Entries are automatically removed if they aren't accessed for 1 hour, which serves to remove empty bin queues
+   * for threads that are no longer reporting distributions.
+   */
+  private final LoadingCache<Long, LinkedList<MinuteBin>> perThreadHistogramBins = Caffeine.newBuilder()
+      .expireAfterAccess(1, TimeUnit.HOURS)
```
Comment (sushantdewan123) [inline java-lib/src/main/java/com/codahale/metrics/WavefrontHistogram.java]: line break after "."
same as above comment.

File: java-client/src/main/java/com/wavefront/integrations/WavefrontDirectSender.java [hunk 11]
```diff
@@ -70,67 +82,104 @@ public void send(String name, double value, @Nullable Long timestamp, String sou
...
-      for (Map.Entry<String, String> entry : pointTags.entrySet()) {
-        sb.append(' ').append(quote).append(escapeQuotes(entry.getKey())).append(quote).
-            append("=").
-            append(quote).append(escapeQuotes(entry.getValue())).append(quote);
+  private void addHistogram(Set<HistogramGranularity> histogramGranularities, @Nullable Long timestamp,
+                            List<Pair<Double, Integer>> distribution, String name, String source,
+                            @Nullable Map<String, String> pointTags) throws IOException {
+    List<String> histograms;
+    try {
+      histograms = WavefrontDataFormat.histogramToStrings(histogramGranularities, timestamp, distribution, name, source,
```
Comment (sushantdewan123) [inline java-client/src/main/java/com/wavefront/integrations/WavefrontDirectSender.java]: line break after name,

**Pull request closed without merging**